CiJenkinsfile

```diff
@@ -0,0 +1,193 @@
+spec:
+  containers:
+  - name: helm
+    image: vfarcic/helm:2.8.2
+    command:
+    - cat
+    tty: true
+  - name: golang
+    image: golang:1.9
+    command: ["cat"]
+    tty: true
+  - name: git
+    image: alpine/git
+    command:
+    - cat
+    tty: true
+  - name: git
+    image: digitalinside/gren:latest
+    command:
+    - cat
+    tty: true    
+  - name: kubectl
+    image: vfarcic/kubectl
+    command: ["cat"]
+    tty: true
+  - name: docker
+    image: docker
+    command:
+    - cat
+    tty: true
+    volumeMounts:
+    - mountPath: /var/run/docker.sock
+      name: docker-sock-volume
+  volumes:
+  - name: docker-sock-volume
+    hostPath:
+      path: /var/run/docker.sock
+      type: File
+"""
+        }
+    }
+
+    parameters {
+        string(name: 'project', defaultValue: 'go-demo-3')
+        string(name: 'image', defaultValue: 'digitalinside/go-demo-3')
+        string(name: 'domain', defaultValue: '192.168.0.10.nip.io')
+        string(name: 'addr', defaultValue: 'go-demo-3.192.168.0.10.nip.io')
+        string(name: 'cmAddr', defaultValue: 'cm.4.192.168.0.10.nip.io')
+        string(name: 'chartVer', defaultValue: '0.0.1')
+    }
+
+    options {
+        timeout(time: 45, unit: 'MINUTES')
+    }
+
+    stages {
+        stage("prep") {
+            steps {
+                script {
+                    //https://issues.jenkins-ci.org/browse/JENKINS-52623 i found issue and waiting it to be fixed, i think carlos is on it now
+                    container('git') {
+                        env.GIT_COMMIT = sh (script: """ git rev-parse HEAD """, returnStdout: true).trim()
+                        env.shortGitCommit = "${env.GIT_COMMIT[0..10]}"
+                        env.BUILD_TAG = ciBuildVersionRead()
+                        echo "build tag set to: ${env.BUILD_TAG}"
+                    }
+
+                    currentBuild.displayName = new SimpleDateFormat("yy.MM.dd").format(new Date()) + "-" + env.BUILD_NUMBER
+                }
+            }
+        }
+
+
+        stage('build') {
+            steps {
+                container('docker') {
+                    //for feature branch
+                    ciK8sBuildImage(params.image, false, env.BUILD_TAG)
+                }
+            }
+        }
+
+        stage("func-test") {
+            steps {
+                container("helm") {
+                    ciK8sUpgradeBeta(params.project, params.domain, env.BUILD_TAG)
+                }
+
+                container("kubectl") {
+                    k8sRolloutBeta(params.project)
+                }
+
+                container("golang") {
+                    ciK8sFuncTestGolang(params.project, params.domain)
+                }
+            }
+
+            post {
+                failure {
+                    container("helm") {
+                        k8sDeleteBeta(params.project)
+                    }
+                }
+            }
+        }
+
+
+        stage ("deploy2uat") {
```

Comment:
If this is uat, where is production? A separate pipeline or in here but yet to be added?